toolchain/check/testdata/impl/no_prelude/impl_cycle.carbon

```diff
@@ -194,3 +194,65 @@ fn F() {
   Compare({} as C, {} as D);
   Compare({} as D, {} as C);
 }
+
+// --- deduce_cycle_without_symbolic.carbon
+library "[[@TEST_NAME]]";
+
+interface First {}
+interface Second {}
+
+// This impl is looked at for "does T impl Second" first, which requires a
+// recursive query in deduction that is the same "does T impl Second", which can
+// land us back looking at this impl again. So we require a way to eliminate
+// this impl without doing deduction in order to avoid the cycle.
+impl forall [T:! Second] T as First {}
+impl forall [T:! type] T as Second {}
```

Comment:
What's the purpose of this impl? The code is still correct without it, and IIUC the potential infinite loop you describe doesn't involve it.

Reply:
Oh, yeah this is redundant with the `impl C as Second` below. I removed the `impl C` one, in both tests.